hamt/hamt.go

```diff
@@ -206,26 +206,41 @@ func (ds *Shard) makeShardValue(lnk *ipld.Link) (*Shard, error) {
 
 // Set sets 'name' = nd in the HAMT
 func (ds *Shard) Set(ctx context.Context, name string, nd ipld.Node) error {
+	_, err := ds.SetAndPrevious(ctx, name, nd)
+	return err
+}
+
+// SetAndPrevious sets a link pointing to the passed node as the value under the
+// name key in this Shard or its children. It also returns the previous link
+// under that name key (if any).
+func (ds *Shard) SetAndPrevious(ctx context.Context, name string, node ipld.Node) (*ipld.Link, error) {
 	hv := &hashBits{b: hash([]byte(name))}
-	err := ds.dserv.Add(ctx, nd)
+	err := ds.dserv.Add(ctx, node)
 	if err != nil {
-		return err
+		return nil, err
 	}
 
-	lnk, err := ipld.MakeLink(nd)
+	lnk, err := ipld.MakeLink(node)
 	if err != nil {
-		return err
+		return nil, err
 	}
 	lnk.Name = ds.linkNamePrefix(0) + name
 
-	return ds.modifyValue(ctx, hv, name, lnk)
+	return ds.setValue(ctx, hv, name, lnk)
 }
 
 // Remove deletes the named entry if it exists. Otherwise, it returns
 // os.ErrNotExist.
 func (ds *Shard) Remove(ctx context.Context, name string) error {
+	_, err := ds.RemoveAndPrevious(ctx, name)
+	return err
+}
+
+// RemoveAndPrevious is similar to the public Remove but also returns the
+// old removed link (if it exists).
+func (ds *Shard) RemoveAndPrevious(ctx context.Context, name string) (*ipld.Link, error) {
```

Comment:
nit: This is usually called `Take`.

Reply:
Agreed, will be done last in https://github.com/ipfs/go-unixfs/issues/105.